File: android/app/src/main/java/dr/achim/sleep_timer/MethodChannelImpl.java
package dr.achim.sleep_timer;

import android.content.Context;
import android.content.Intent;
import android.content.pm.ApplicationInfo;
import android.content.pm.PackageInfo;
import android.content.pm.PackageManager;
import android.content.pm.ResolveInfo;
import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.drawable.Drawable;
import android.net.Uri;
import android.os.Build;
import android.provider.AlarmClock;
import android.provider.MediaStore;
import android.util.Base64;

import androidx.annotation.NonNull;
import androidx.core.app.NotificationManagerCompat;

import org.jetbrains.annotations.NotNull;

import java.io.ByteArrayOutputStream;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

import dr.achim.sleep_timer.Messages.*;
import io.flutter.Log;

public class MethodChannelImpl implements HostTimerApi {
    private static final String TAG = MethodChannelImpl.class.toString();
    private final Context context;

    public MethodChannelImpl(Context context) {
        this.context = context;
    }

    @Override
    public void init(InitializationRequest arg) {
        CallbackHelper.setHandle(context, arg.getCallbackHandle());
    }

    private void _startForegroundService(final TimeNotificationRequest arg) {
        final Intent intent = new Intent(context, AlarmService.class);
        intent.setAction(AlarmService.ACTION_START);
        intent.putExtra(NotificationReceiver.KEY_SHOW_NOTIFICATION, arg.toMap());

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            context.startForegroundService(intent);
        } else {
            context.startService(intent);
        }
    }

    private void _stopForegroundService() {
        if(AlarmService.isRunning()) {
            final Intent intent = new Intent(context, AlarmService.class);
            context.stopService(intent);
        }
    }

    @Override
    public NotificationResponse showRunningNotification(TimeNotificationRequest arg) {
        final NotificationResponse response = new NotificationResponse();
        final String timerId = arg.getTimerId();
        if(timerId == null) {
            response.setSuccess(false);
            return response;
        }

        Log.wtf(TAG, "Request to show running notification for timer with id " + timerId);

        _startForegroundService(arg);
        // Notification will be triggered by broadcast when foreground service is ready

        response.setTimerId(timerId);
        response.setSuccess(true);
        return response;
    }

    @Override
    public NotificationResponse showPausingNotification(TimeNotificationRequest arg) {
        final NotificationResponse response = new NotificationResponse();
        final String timerId = arg.getTimerId();
        if(timerId == null) {
            response.setSuccess(false);
            return response;
        }

        Log.wtf(TAG, "Request to show pause notification for timer with id " + timerId);

        _stopForegroundService();

        final Intent intent = new Intent(context, NotificationReceiver.class);
        intent.setAction(NotificationReceiver.ACTION_PAUSE_NOTIFICATION);
        intent.putExtra(NotificationReceiver.KEY_PAUSE_NOTIFICATION, arg.toMap());
        context.sendBroadcast(intent);

        response.setTimerId(timerId);
        response.setSuccess(true);
        return response;
    }

    @Override
    public NotificationResponse showElapsedNotification(NotificationRequest arg) {
        final NotificationResponse response = new NotificationResponse();
        final String timerId = arg.getTimerId();
        if(timerId == null) {
            response.setSuccess(false);
            return response;
        }

        Log.wtf(TAG, "Request to show elapsed notification for timer with id " + timerId);

        _stopForegroundService();

        final Intent intent = new Intent(context, NotificationReceiver.class);
        intent.setAction(NotificationReceiver.ACTION_ELAPSED_NOTIFICATION);
        intent.putExtra(NotificationReceiver.KEY_ELAPSED_NOTIFICATION, arg.toMap());
        context.sendBroadcast(intent);

        response.setTimerId(timerId);
        response.setSuccess(true);
        return response;
    }

    @Override
    public CancelResponse cancelTimer(CancelRequest arg) {
        final CancelResponse response = new CancelResponse();
        final String timerId = arg.getTimerId();
        if(timerId == null) {
            response.setSuccess(false);
            return response;
        }

        Log.wtf(TAG, "Request to cancel notification for timer with id " + timerId);

        _stopForegroundService();
        NotificationManagerCompat.from(context).cancel(NotificationReceiver.NOTIFICATION_ID);

        response.setTimerId(timerId);
        response.setSuccess(true);
        return response;
    }

    @Override
    public InstalledAppsResponse getInstalledPlayerApps() {
        final InstalledAppsResponse response = new InstalledAppsResponse();

        final PackageManager manager = context.getPackageManager();
        Intent intentAudio = new Intent(Intent.ACTION_VIEW);
        Uri uriAudio = Uri.withAppendedPath(MediaStore.Audio.Media.INTERNAL_CONTENT_URI, "1");
        intentAudio.setDataAndType(uriAudio, "audio/*");
        List<ResolveInfo> audioList = manager.queryIntentActivities(intentAudio, 0);

        Intent intentVideo = new Intent(Intent.ACTION_VIEW);
        Uri uriVideo = Uri.withAppendedPath(MediaStore.Video.Media.INTERNAL_CONTENT_URI, "1");
        intentVideo.setDataAndType(uriVideo, "video/*");
        List<ResolveInfo> videoList = manager.queryIntentActivities(intentVideo, 0);

        Intent intentMedia = new Intent(Intent.ACTION_MEDIA_BUTTON);
        List<ResolveInfo> mediaList = manager.queryBroadcastReceivers(intentMedia, 0);

        final List<Messages.Package> apps = new ArrayList<>();
        apps.addAll(getAppsFromResolveInfo(audioList));
        apps.addAll(getAppsFromResolveInfo(videoList));
        apps.addAll(getAppsFromResolveInfo(mediaList));


            List<String> packageNames = new ArrayList<>();
            packageNames.add("com.netflix.mediaclient");
            packageNames.add("com.google.android.apps.podcasts");

            for (String packageName : packageNames) {
                try {
                    final PackageInfo myPackage = manager.getPackageInfo(packageName, 0);
                    final ApplicationInfo applicationInfo = myPackage.applicationInfo;
                    if (applicationInfo != null && applicationInfo.enabled) {
                        final Messages.Package app = getAppFromApplicationInfo(applicationInfo);
                        apps.add(app);
                    }
                } catch (PackageManager.NameNotFoundException ignored) {}
            }

        final ArrayList<HashMap> distinctApps = getDistinctAppList(apps);
        response.setApps(distinctApps);
        return response;
    }

    @Override
    public InstalledAppsResponse getInstalledAlarmApps() {
        final InstalledAppsResponse response = new InstalledAppsResponse();

        final PackageManager manager = context.getPackageManager();

        Intent showAlarmsIntent = new Intent(AlarmClock.ACTION_SHOW_ALARMS);
        List<ResolveInfo> alarmList = manager.queryIntentActivities(showAlarmsIntent, 0);

        final List<Messages.Package> apps = getAppsFromResolveInfo(alarmList);
        final ArrayList<HashMap> distinctApps = getDistinctAppList(apps);
        response.setApps(distinctApps);
        return response;
    }

    @NotNull
    private List<Messages.Package> getAppsFromResolveInfo(List<ResolveInfo> infoList) {
        List<Messages.Package> apps = new ArrayList<>();
        for(ResolveInfo info : infoList) {
            final ApplicationInfo applicationInfo = info.activityInfo.applicationInfo;
            final Messages.Package app = getAppFromApplicationInfo(applicationInfo);
            apps.add(app);
        }
        return apps;
    }

    @NotNull
    private Messages.Package getAppFromApplicationInfo(final ApplicationInfo applicationInfo) {
        final PackageManager manager = context.getPackageManager();

        // https://stackoverflow.com/questions/51368075/how-can-i-get-android-drawables-in-flutter
        final Bitmap bitmap = getBitmapFromDrawable(applicationInfo.loadIcon(manager));
        final String encoded = getBase64FromBitmap(bitmap);

        Messages.Package app = new Messages.Package();
        app.setPackageName(applicationInfo.packageName);
        app.setTitle(applicationInfo.loadLabel(manager).toString());
        app.setIcon(encoded);

        return app;
    }

    /**
     * Ensure that there are no duplicates.
     */
    @NotNull
    private ArrayList<HashMap> getDistinctAppList(final List<Messages.Package> apps) {
        Set<String> distinctAppSet = new HashSet<>();
        final ArrayList<HashMap> distinctApps = new ArrayList<>();
        for (Messages.Package app : apps) {
            if(!distinctAppSet.contains(app.getPackageName())) {
                distinctAppSet.add(app.getPackageName());
                distinctApps.add(app.toMap());
            }
        }
        return distinctApps;
    }

    @Override
    public void launchApp(final LaunchAppRequest request) {
        final PackageManager manager = context.getPackageManager();
        final Intent intent = manager.getLaunchIntentForPackage(request.getPackageName());
        context.startActivity(intent);
    }

    @Override
    public Messages.Package dummyApp() {
        return null;
    }

    // https://stackoverflow.com/questions/44447056/convert-adaptiveicondrawable-to-bitmap-in-android-o-preview
    @NonNull
    private Bitmap getBitmapFromDrawable(@NonNull Drawable drawable) {
        final Bitmap bitmap = Bitmap.createBitmap(drawable.getIntrinsicWidth(), drawable.getIntrinsicHeight(), Bitmap.Config.ARGB_8888);
        final Canvas canvas = new Canvas(bitmap);
        drawable.setBounds(0, 0, canvas.getWidth(), canvas.getHeight());
        drawable.draw(canvas);
        return bitmap;
    }

    // https://stackoverflow.com/questions/9224056/android-bitmap-to-base64-string/9224180#9224180
    @NonNull
    private String getBase64FromBitmap(@NonNull Bitmap bitmap) {
        final ByteArrayOutputStream byteArrayOutputStream = new ByteArrayOutputStream();
        bitmap.compress(Bitmap.CompressFormat.PNG, 100, byteArrayOutputStream);
        final byte[] byteArray = byteArrayOutputStream .toByteArray();
        return Base64.encodeToString(byteArray, Base64.NO_WRAP);
    }

}

File: android/app/src/main/kotlin/dr/achim/sleep_timer/MethodChannelImpl.kt
package dr.achim.sleep_timer

import android.content.Context
import android.content.Intent
import android.content.pm.ApplicationInfo
import android.content.pm.PackageManager
import android.content.pm.ResolveInfo
import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.drawable.Drawable
import android.net.Uri
import android.provider.AlarmClock
import android.provider.MediaStore
import android.util.Base64
import androidx.core.app.NotificationManagerCompat
import androidx.core.content.ContextCompat
import dr.achim.sleep_timer.Messages.*
import io.flutter.Log
import java.io.ByteArrayOutputStream
import java.util.*


class MethodChannelImpl(private val context: Context) : HostTimerApi {

    companion object {
        private val TAG = MethodChannelImpl::class.java.toString()

        private val audioPlayerPackages = listOf(
                "com.spotify.music",
                "com.google.android.youtube",
                "deezer.android.app",
                "com.apple.android.music",
                "com.amazon.mp3",
                "com.pandora.android",
                "com.rhapsody",
                "com.aspiro.tidal",
                "com.soundcloud.android",
                "com.google.android.apps.podcasts",
                "com.sec.android.app.music",
                "com.google.android.apps.youtube.music",
        )

        private val videoPlayerPackages = listOf(
                "com.netflix.mediaclient",
                "com.disney.disneyplus",
                "com.amazon.avod.thirdpartyclient",
                "com.disney.disneyplus",
                "de.prosiebensat1digital.seventv",
                "de.sky.bw",
                "de.sky.online",
                "com.dazn",
                "com.plexapp.android",
                "org.xbmc.kodi",
                "com.vanced.android.youtube",
        )

        private val excludePackages = listOf(
                "com.google.android.apps.docs",
                "com.google.android.apps.photos",
        )
    }

    private fun startForegroundService(arg: RunningNotificationRequest) {
        val intent = Intent(context, AlarmService::class.java).apply {
            action = AlarmService.ACTION_START
            putExtra(NotificationReceiver.KEY_SHOW_NOTIFICATION, arg.toList())
        }

        ContextCompat.startForegroundService(context, intent)
    }

    private fun stopForegroundService() {
        if (AlarmService.isRunning) {
            val intent = Intent(context, AlarmService::class.java)
            context.stopService(intent)
        }
    }

    override fun showRunningNotification(arg: RunningNotificationRequest): NotificationResponse {
        stopForegroundService()
        val response = NotificationResponse().apply {
            timerId = arg.timerId
            success = true
        }

        if (response.timerId == null) {
            return response.apply { success = false }
        }

        Log.d(TAG, "Request to show running notification for timer with id ${response.timerId}")
        startForegroundService(arg)
        // Notification will be triggered by broadcast when foreground service is ready
        return response
    }

    override fun showPausingNotification(arg: TimeNotificationRequest): NotificationResponse {
        val response = NotificationResponse().apply {
            timerId = arg.timerId
            success = true
        }

        if (response.timerId == null) {
            return response.apply { success = false }
        }

        Log.d(TAG, "Request to show pause notification for timer with id ${response.timerId}")
        stopForegroundService()
        val intent = Intent(context, NotificationReceiver::class.java).apply {
            action = NotificationReceiver.ACTION_PAUSE_NOTIFICATION
            putExtra(NotificationReceiver.KEY_PAUSE_NOTIFICATION, arg.toList())
        }
        context.sendBroadcast(intent)

        return response
    }

    override fun showElapsedNotification(arg: NotificationRequest): NotificationResponse {
        val response = NotificationResponse().apply {
            timerId = arg.timerId
            success = true
        }

        if (response.timerId == null) {
            return response.apply { success = false }
        }

        Log.d(TAG, "Request to show elapsed notification for timer with id ${response.timerId}")
        stopForegroundService()
        val intent = Intent(context, NotificationReceiver::class.java).apply {
            action = NotificationReceiver.ACTION_ELAPSED_NOTIFICATION
            putExtra(NotificationReceiver.KEY_ELAPSED_NOTIFICATION, arg.toList())
        }
        context.sendBroadcast(intent)

        return response
    }

    override fun cancelTimer(arg: CancelRequest): CancelResponse {
        val response = CancelResponse().apply {
            timerId = arg.timerId
            success = true
        }

        if (response.timerId == null) {
            return response.apply { success = false }
        }
        Log.d(TAG, "Request to cancel notification for timer with id ${response.timerId}")
        stopForegroundService()
        NotificationManagerCompat.from(context).cancel(NotificationReceiver.NOTIFICATION_ID)

        return response
    }

    override fun toggleExtendByShake(arg: ToggleRequest) {
        Log.d(TAG, "Request to toggle extend by shake: enable=${arg.enable}")
        if (AlarmService.isRunning) {
            val intent = Intent(context, AlarmService::class.java).apply {
                action = AlarmService.ACTION_TOGGLE_EXTEND_BY_SHAKE
                putExtra(AlarmService.KEY_ENABLE_EXTEND_BY_SHAKE, arg.enable)
            }

            context.startService(intent)
        }
    }

    override fun getInstalledPlayerApps(): InstalledAppsResponse {
        val manager = context.packageManager

        val uriAudio = Uri.withAppendedPath(MediaStore.Audio.Media.INTERNAL_CONTENT_URI, "1")
        val intentAudio = Intent(Intent.ACTION_VIEW).apply {
            setDataAndType(uriAudio, "audio/*")
        }
        val audioList = manager.queryIntentActivities(intentAudio, 0)

        val uriVideo = Uri.withAppendedPath(MediaStore.Video.Media.INTERNAL_CONTENT_URI, "1")
        val intentVideo = Intent(Intent.ACTION_VIEW).apply {
            setDataAndType(uriVideo, "video/*")
        }
        val videoList = manager.queryIntentActivities(intentVideo, 0)

        val intentMedia = Intent(Intent.ACTION_MEDIA_BUTTON)
        val mediaList = manager.queryBroadcastReceivers(intentMedia, 0)

        val apps: MutableSet<Package> = getAppsFromResolveInfo(audioList)
                .union(getAppsFromResolveInfo(videoList))
                .union(getAppsFromResolveInfo(mediaList))
                .filterNot { excludePackages.contains(it.packageName) }
                .toMutableSet()

        // Query by packageName
        val packageNames = audioPlayerPackages + videoPlayerPackages
        for (packageName in packageNames) {
            try {
                val myPackage = manager.getPackageInfo(packageName, 0)
                val applicationInfo = myPackage.applicationInfo
                if (applicationInfo != null && applicationInfo.enabled) {
                    val app = getAppFromApplicationInfo(applicationInfo)
                    apps.add(app)
                }
            } catch (ignored: PackageManager.NameNotFoundException) {
                Log.d(TAG, "$packageName not found")
            }
        }

        return InstalledAppsResponse().apply {
            this.apps = apps.distinctBy { it.packageName }
        }
    }

    override fun getInstalledAlarmApps(): InstalledAppsResponse {
        val showAlarmsIntent = Intent(AlarmClock.ACTION_SHOW_ALARMS)
        val alarmList = context.packageManager.queryIntentActivities(showAlarmsIntent, 0)
        val apps = getAppsFromResolveInfo(alarmList).distinctBy { it.packageName }

        return InstalledAppsResponse().apply {
            this.apps = apps
        }
    }

    private fun getAppsFromResolveInfo(infoList: List<ResolveInfo>): List<Package> {
        val apps: MutableList<Package> = ArrayList()
        for (info in infoList) {
            val applicationInfo = info.activityInfo.applicationInfo
            val app = getAppFromApplicationInfo(applicationInfo)
            apps.add(app)
        }
        return apps
    }

    private fun getAppFromApplicationInfo(applicationInfo: ApplicationInfo): Package {
        val manager = context.packageManager

        // https://stackoverflow.com/questions/51368075/how-can-i-get-android-drawables-in-flutter
        val bitmap = getBitmapFromDrawable(applicationInfo.loadIcon(manager))
        val encoded = getBase64FromBitmap(bitmap)
        return Package().apply {
            packageName = applicationInfo.packageName
            title = applicationInfo.loadLabel(manager).toString()
            icon = encoded
        }
    }

    override fun launchApp(request: LaunchAppRequest) {
        val manager = context.packageManager
        val intent = manager.getLaunchIntentForPackage(request.packageName!!)
        context.startActivity(intent)
    }

    // https://stackoverflow.com/questions/44447056/convert-adaptiveicondrawable-to-bitmap-in-android-o-preview
    private fun getBitmapFromDrawable(drawable: Drawable): Bitmap {
        val bitmap = Bitmap.createBitmap(drawable.intrinsicWidth, drawable.intrinsicHeight, Bitmap.Config.ARGB_8888)
        val canvas = Canvas(bitmap)
        drawable.setBounds(0, 0, canvas.width, canvas.height)
        drawable.draw(canvas)
        return bitmap
    }

    // https://stackoverflow.com/questions/9224056/android-bitmap-to-base64-string/9224180#9224180
    private fun getBase64FromBitmap(bitmap: Bitmap): String {
        val byteArrayOutputStream = ByteArrayOutputStream()
        bitmap.compress(Bitmap.CompressFormat.PNG, 100, byteArrayOutputStream)
        val byteArray = byteArrayOutputStream.toByteArray()
        return Base64.encodeToString(byteArray, Base64.NO_WRAP)
    }

}